Fix flax whisper tokenizer bug

hannan72 opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes Bug when using whisper tokenizer for flax whisper model, according to the issue #32936


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

@ArthurZucker 
@sanchit-gandhi

hannan72 committed:
Update tokenization_whisper.py

Fix issue with flax whisper model

hannan72 committed:
Update tokenization_whisper_fast.py

Fix issue with flax whisper model

amyeroberts commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -849,7 +849,7 @@ def _strip_prompt(self, token_ids: List[int], prompt_token_id: int, decoder_star
         # handle case of empty token_ids for decoding with timestamps.
         # at this point token_ids is a list, so it is safe to use if not check.
-        if not token_ids:
+        if token_ids is None or len(token_ids) == 0:
```
This doesn't match with the comment above -- which indicates unexpected behaviour in the `_convert_to_list` method. 

The linked issue indicates a problem with checking `not` on np arrays, but this doesn't correspond to the `None` check here

hannan72 replied:
Good points!
I also updated the comment
Please take a look again

amyeroberts replied:
The comment still doesn't make sense wrt the change. If `token_ids` really is a list, then checking `not token_ids` should be safe

Hannan-Komari replied:
@amyeroberts The problem is that `token_ids` is not a list for flax models, but it is a jax array. So that it raises error for `not token_ids`

hannan72 replied:
@amyeroberts The problem is that token_ids is not a list for flax models, but it is a jax array. So that it raises error for not token_ids

hannan72 replied:
@amyeroberts Do you have any other suggestion for resolving the issue for flax models?

amyeroberts replied:
The main thing is that the function names and comments should be consistent with the logic. So in the comment above it says the object is a list, which appears not to be true. In addition, the issue indicates `_convert_to_list` isn't actually converting to a list. So the issues to resolve is why isn't `_convert_to_list` converting to list? Should the method be updated or changed?

hannan72 replied:
@amyeroberts You're right! So we should update `_convert_to_list` method inorder to convert jax arrays to list as long as numpy, torch and tf arrays.
I updated the PR, reverted the changes in checking `token_ids` again with `not` and add a couple of lines in `_convert_to_list` to convert jax array to list.
Is it OK now?

amyeroberts reviewed (commented):
Thanks for opening a PR @hannan72! 

Please make sure when opening a PR to only tag a minimal subset of the most relevant people. This ensures the PR is reviewed quickly (no by-stander effect) and is good practice and it keeps the number of notifications for everyone in check.

hannan72 committed:
Update tokenization_whisper.py

just check len of token_ids

hannan72 committed:
Update tokenization_whisper_fast.py

just use len of token_ids

hannan72 commented:
Any other points @amyeroberts ?

hannan72 committed:
Update tokenization_whisper_fast.py and revert changes in _strip_prompt and add support to jax arrays in _convert_to_list

hannan72 committed:
Update tokenization_whisper.py and revert changes in _strip_prompt and add support to jax arrays in _convert_to_list

hannan72 commented:
Is it ready to merge  @amyeroberts?

amyeroberts commented:
@hannan72 Thanks for iterating - change now looks OK - final thing is to add a test, which would fail on current `main` but passes with this fix

hannan72 committed:
Update test_tokenization_whisper.py to add test for _convert_to_list method

hannan72 committed:
Update test_tokenization_whisper.py to fix code style issues

hannan72 committed:
Fix code style

hannan72 committed:
Fix code check again

hannan72 committed:
Update test_tokenization)whisper.py to Improve code style

hannan72 commented:
> @hannan72 Thanks for iterating - change now looks OK - final thing is to add a test, which would fail on current `main` but passes with this fix

@amyeroberts I added a test to `tests/models/whisper/test_tokenization_whisper.py` for `_convert_to_list` method that I modified.
There is one issue about it. This test is running on `huggingface/transformers-torch-light` container which 'jax' is not installed on it, while `jax` module is needed to test the updated `_convert_to_list` method.
What should I do for that? How can I change the docker image to `huggingface/transformers-torch-jax-light` for this test?

hannan72 committed:
Update test_tokenization_whisper.py to run each of jax, tf and flax modules if available

hannan72 commented:
@amyeroberts Test codes has been added and passed by automatic tests. Code you please do the final review?

amyeroberts commented on tests/models/whisper/test_tokenization_whisper.py:
```diff
@@ -574,3 +617,33 @@ def test_offset_decoding(self):
         output = multilingual_tokenizer.decode(INPUT_TOKENS, output_offsets=True)["offsets"]
         self.assertEqual(output, [])
+
+    def test_convert_to_list(self):
+        test_list = [[1, 2, 3], [4, 5, 6]]
+
+        # Test with an already converted list
+        assert WhisperTokenizer._convert_to_list(test_list) == test_list
+        assert WhisperTokenizerFast._convert_to_list(test_list) == test_list
+
+        # Test with a numpy array
+        np_array = np.array(test_list)
+        assert WhisperTokenizer._convert_to_list(np_array) == test_list
+        assert WhisperTokenizerFast._convert_to_list(np_array) == test_list
+
+        if is_torch_available():
+            # Test with a PyTorch tensor
+            torch_tensor = torch.tensor(test_list)
+            assert WhisperTokenizer._convert_to_list(torch_tensor) == test_list
+            assert WhisperTokenizerFast._convert_to_list(torch_tensor) == test_list
+
+        if is_tf_available():
+            # Test with a TensorFlow tensor
+            tf_tensor = tf.constant(test_list)
+            assert WhisperTokenizer._convert_to_list(tf_tensor) == test_list
+            assert WhisperTokenizerFast._convert_to_list(tf_tensor) == test_list
+
+        if is_flax_available():
+            # Test with a JAX array
+            jax_array = jnp.array(test_list)
+            assert WhisperTokenizer._convert_to_list(jax_array) == test_list
+            assert WhisperTokenizerFast._convert_to_list(jax_array) == test_list
```
* tests should be controlled with the `@require_xxx` decorator
* we should use unittest test cases to test - ListEqual will give us a richer output for debugging 
* we should split up the tests to be more atomic

```suggestion
    def test_convert_to_list_np(self):
        test_list = [[1, 2, 3], [4, 5, 6]]

        # Test with an already converted list  
        self.assertListEqual(WhisperTokenizer._convert_to_list(test_list), test_list)
        self.assertListEqual(WhisperTokenizerFast._convert_to_list(test_list), test_list)

        # Test with an already converted list  
        np_array = np.array(test_list)
        self.assertListEqual(WhisperTokenizer._convert_to_list(test_list), test_list)
        self.assertListEqual(WhisperTokenizerFast._convert_to_list(test_list), test_list)

    @require_tf
    def test_convert_to_list_tf(self):
        test_list = [[1, 2, 3], [4, 5, 6]]

        tf_tensor = tf.constant(test_list)
        self.assertListEqual(WhisperTokenizer._convert_to_list(test_list), test_list)
        self.assertListEqual(WhisperTokenizerFast._convert_to_list(test_list), test_list)

    @require_flax
    def test_convert_to_list_jax(self):
        test_list = [[1, 2, 3], [4, 5, 6]]

        jax_array = jnp.array(test_list) 
        self.assertListEqual(WhisperTokenizer._convert_to_list(test_list), test_list)
        self.assertListEqual(WhisperTokenizerFast._convert_to_list(test_list), test_list)

    @require_torch
    def test_convert_to_list_pt(self):
        test_list = [[1, 2, 3], [4, 5, 6]]

        torch_tensor = torch.tensor(test_list)
        self.assertListEqual(WhisperTokenizer._convert_to_list(test_list), test_list)
        self.assertListEqual(WhisperTokenizerFast._convert_to_list(test_list), test_list)
```

amyeroberts reviewed (commented):
Thanks for iterating and adding tests! 

General comment that unrelated formatting changes should be removed from the diff. Once the tests are split up we should be good to go

hannan72 committed:
Update tests/models/whisper/test_tokenization_whisper.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

hannan72 committed:
Update test_tokenization_whisper.py and use require_xxx decorators instead of `is_xxx_available()` method

hannan72 commented:
Thanks @amyeroberts for your suggestion! It is applied and tests are split up! Just needs your approval!

hannan72 commented:
> Thanks for iterating and adding tests!
> 
> General comment that unrelated formatting changes should be removed from the diff. Once the tests are split up we should be good to go

Anything else?

amyeroberts commented on tests/models/whisper/test_tokenization_whisper.py:
```diff
@@ -204,11 +217,21 @@ def test_skip_special_tokens_skips_prompt_ids(self):
         # fmt: on
         expected_with_special_tokens = "<|startofprev|> Mr. Quilter<|startoftranscript|><|en|><|transcribe|><|notimestamps|> On the general principles of art, Mr. Quilter writes with equal lucidity.<|endoftext|>"
         expected_without_special_tokens = " On the general principles of art, Mr. Quilter writes with equal lucidity."
-        self.assertEqual(tokenizer.decode(encoded_input, skip_special_tokens=False), expected_with_special_tokens)
```
Can you remove all these changes which shouldn't be applied (our line length is 120 and this is a formatting change unrelated to the PR)

hannan72 replied:
@amyeroberts All unrelated changes have been reverted. Now is it the proper time for merging the PR?

hannan72 committed:
Revert the changes automatically applied by formatter and was unrelated to PR

hannan72 committed:
Format for minimal changes

hannan72 commented:
@amyeroberts All changes has been made. I'd appreciate it if you merge the PR!

amyeroberts reviewed (approved):
Great - thanks for fixing and iterating on this!
